### PR TITLE
Improve map, optimize and StreamingDataset

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR improves the API toward immutability.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18912.org.readthedocs.build/en/18912/

cc @borda

### CODE REVIEW

File: src/lightning/data/streaming/dataset.py [hunk 1]
```diff
@@ -53,12 +74,19 @@ def __init__(
...
+        if cache_dir:
+            input_dir.path = cache_dir
+
+        self.cache = Cache(input_dir=input_dir, item_loader=item_loader, chunk_bytes=1)
 
         self.cache._reader._try_load_config()
 
         if not self.cache.filled:
-            raise ValueError(f"The provided dataset `{name}` isn't filled up.")
+            raise ValueError(f"The provided dataset `{input_dir}` isn't filled up.")
```
Comment (lantiga) [inline src/lightning/data/streaming/dataset.py:89]: This is a pretty cryptic error message IMO. It's probably not likely to be happen for actual users but a bit more context could help.

File: src/lightning/data/streaming/data_processor.py [hunk 2]
```diff
@@ -116,16 +110,19 @@ def _download_data_target(
         index, paths = r
 
         # 5. Check whether all the files are already downloaded
-        if all(os.path.exists(p.replace(input_dir, cache_dir) if input_dir else p) for p in paths):
+        if all(os.path.exists(p.replace(input_dir.path, cache_dir) if input_dir else p) for p in paths):
```
Comment (lantiga) [inline src/lightning/data/streaming/data_processor.py:118]: They might have been downloaded but not completely, is there anything we can do to alleviate?

**Review by dmitsf — APPROVED** [review]

File: src/lightning/data/streaming/functions.py [hunk 3]
```diff
@@ -168,7 +177,7 @@ def optimize(
         fast_dev_run: Whether to use process only a sub part of the inputs
         num_nodes: When doing remote execution, the number of nodes to use.
         machine: When doing remote execution, the machine to use.
-        num_downloaders: The number of downloaders per worker.
+        num_downloaders: The number of downloaders to use.
```
Comment (lantiga) [inline src/lightning/data/streaming/functions.py:180]: Maybe "number of downloaders per machine" would be clearer?

**Review by lantiga — APPROVED** [review]
Looks great, a few comments

File: src/lightning/data/streaming/cache.py [hunk 4]
```diff
@@ -13,32 +13,33 @@
...
-    from lightning_cloud.resolver import _find_remote_dir, _try_create_cache_dir
-
 logger = logging.Logger(__name__)
 
+if _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_47:
+    from lightning_cloud.resolver import Dir, _resolve_dir
+else:
+
+    class Dir:
+        pass
```
Comment (awaelchli) [inline src/lightning/data/streaming/cache.py:36]: Can't we set lightning-cloud 0.5.47 as the minimum version?

File: src/lightning/data/streaming/cache.py [hunk 5]
```diff
@@ -63,25 +62,20 @@ def __init__(
...
-
-            if cache_dir:
-                os.makedirs(cache_dir, exist_ok=True)
-
-            self._cache_dir = cache_dir
-
-        self._writer = BinaryWriter(cache_dir, chunk_size=chunk_size, chunk_bytes=chunk_bytes, compression=compression)
-        self._reader = BinaryReader(cache_dir, remote_dir=remote_dir, compression=compression, item_loader=item_loader)
+        if not _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_47:
+            raise ModuleNotFoundError("Lightning Cloud 0.5.47 or higher is required to use the cache.")
```
Comment (awaelchli) [inline src/lightning/data/streaming/cache.py:66]: see comment above

File: src/lightning/data/streaming/dataset.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:89]: Done !

File: src/lightning/data/streaming/data_processor.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/data_processor.py:118]: We clean up the cache folder every time you start the script.

File: src/lightning/data/streaming/cache.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/data/streaming/cache.py:36]: done !

File: src/lightning/data/streaming/cache.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/streaming/cache.py:66]: done !

File: src/lightning/data/streaming/downloader.py [hunk 6]
```diff
@@ -66,14 +66,15 @@ class LocalDownloader(Downloader):
     def download_file(cls, remote_filepath: str, local_filepath: str) -> None:
         if not os.path.exists(remote_filepath):
             raise FileNotFoundError("The provided remote_path doesn't exist: {remote_path}")
-        shutil.copy(remote_filepath, local_filepath)
+        if remote_filepath != local_filepath:
+            shutil.copy(remote_filepath, local_filepath)
```
Comment (awaelchli) [inline src/lightning/data/streaming/downloader.py:70]: Is this fixing a bug? We should have a test in test_dataloaders() for these edge cases I think

Comment (tchaton) [reply src/lightning/data/streaming/downloader.py:70]: This is more to make sure we can do local testing.

File: src/lightning/data/streaming/functions.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/functions.py]: Reverted the change. This would be removed in further iteration by using some kinda of auto-tuning.

File: requirements/app/app.txt [hunk 7]
```diff
@@ -1,6 +1,6 @@
-lightning-cloud ==0.5.46  # Must be pinned to ensure compatibility
+lightning-cloud == 0.5.48  # Must be pinned to ensure compatibility
 packaging
-typing-extensions >=4.0.0, <4.8.0
+typing-extensions >=4.4.0, <4.8.0
```
Comment (Borda) [inline requirements/app/app.txt:3]: why here, this was another PR

**Review by awaelchli — APPROVED** [review]
Haven't looked thoroughly, I struggle to keep up with lightning.data at the moment. Apprv to unblock

File: requirements/app/app.txt (hunk 7 above)
Comment (tchaton) [reply requirements/app/app.txt:3]: CI was breaking last night. This seemed this was required.
